### PR TITLE
One hot encoded labels support

### PR DESCRIPTION
**Pull request by swetha097**
This PR adds support for the one hot encoded labels in c++ and python

### CODE REVIEW

File: rocAL_pybind/amd/rocal/pipeline.py [hunk 1]
```diff
@@ -153,22 +152,15 @@ def copyToExternalTensor(self, array,  multiplier, offset, reverse_channels, ten
...
-    NOTE: getOneHotEncodedLabels is not defined. This function does not have a cpp equivalent. Bring back once cupy/fucntion is fixed
-    def get_one_hot_encoded_labels(self, array, device):
+    def get_one_hot_encoded_labels(self, array_ptr, device):
         if device == "cpu":
-            if (isinstance(array, np.ndarray)):
-                b.getOneHotEncodedLabels(self._handle, array.ctypes.data_as(
-                    ctypes.c_void_p), self._num_classes, 0)
-            else:  # torch tensor
-                return b.getOneHotEncodedLabels(self._handle, ctypes.c_void_p(array.data_ptr()), self._num_classes, 0)
+            b.getOneHotEncodedLabels(self._handle, array_ptr, self._num_classes, types.HOST_MEMORY)
```
Comment (rrawther) [inline rocAL_pybind/amd/rocal/pipeline.py:157]: the same function is called with same parameters for both cpu and gpu. In this case, if else condition can be removed

File: rocAL_pybind/amd/rocal/pipeline.py [hunk 2]
```diff
@@ -153,22 +152,15 @@ def copyToExternalTensor(self, array,  multiplier, offset, reverse_channels, ten
...
         else:
-            if (isinstance(array, cp.ndarray)):
-                b.getCupyOneHotEncodedLabels(
-                    self._handle, array.data.ptr, self._num_classes, 1)
-            else:  # torch tensor
-                return b.getOneHotEncodedLabels(self._handle, ctypes.c_void_p(array.data_ptr()), self._num_classes, 1)
-    """
+            b.getOneHotEncodedLabels(self._handle, array_ptr, self._num_classes, types.DEVICE_MEMORY)
+    
+    def get_one_hot_encoded_labels_cupy(self, array_ptr):
```
Comment (rrawther) [inline rocAL_pybind/amd/rocal/pipeline.py:161]: get_one_hot_encoded_labels_cupy function

File: rocAL_pybind/amd/rocal/plugin/generic.py [hunk 3]
```diff
@@ -157,8 +157,13 @@ def __next__(self):
...
         if self.loader._name == "labelReader":
             if self.loader._one_hot_encoding == True:
-                self.loader.get_one_hot_encoded_labels(
-                    self.labels, self.device)
+                if self.device == "cpu":
+                    self.loader.get_one_hot_encoded_labels(
+                        self.labels.ctypes.data_as(
+                    ctypes.c_void_p), self.device)
+                else:
+                    self.loader.get_one_hot_encoded_labels_cupy(
```
Comment (rrawther) [inline rocAL_pybind/amd/rocal/plugin/generic.py:165]: change to more generic name like gpu. Why this function needs to be different for cpu Vs gpu as we are passing a pointer to labels in both cases

File: rocAL_pybind/amd/rocal/plugin/tf.py [hunk 4]
```diff
@@ -183,16 +184,18 @@ def __next__(self):
                     self.labels = np.zeros(
                         (self.bs) * (self.loader._num_classes), dtype="int32")
                     self.loader.get_one_hot_encoded_labels(
-                        self.labels, device="cpu")
+                        self.labels.ctypes.data_as(
```
Comment (rrawther) [inline rocAL_pybind/amd/rocal/plugin/tf.py:187]: I think it is better to pass (void *) for both cpu/gpu use cases. The low level function can interpret it as gpu mem pointer

File: rocAL_pybind/rocal_pybind.cpp [hunk 5]
```diff
@@ -129,6 +129,20 @@ py::object wrapperRocalExternalSourceFeedInput(
     return py::cast<py::none>(Py_None);
 }
 
+    py::object wrapper_one_hot_label_copy(RocalContext context, py::object p , unsigned num_of_classes, RocalOutputMemType dest_mem_type) {
+        auto ptr = ctypes_void_ptr(p);
```
Comment (rrawther) [inline rocAL_pybind/rocal_pybind.cpp:133]: please pass as (void*) from high level and remove redundant function

**Review by rrawther — CHANGES_REQUESTED** [review]
requested few changes

File: rocAL_pybind/amd/rocal/pipeline.py (hunk 1 above)
Comment (swetha097) [reply rocAL_pybind/amd/rocal/pipeline.py]: Merged CPU and GPU functionalities for one hot labels

File: rocAL_pybind/amd/rocal/pipeline.py (hunk 2 above)
Comment (swetha097) [reply rocAL_pybind/amd/rocal/pipeline.py]: done. Merged it to a single call

File: rocAL_pybind/rocal_pybind.cpp (hunk 5 above)
Comment (swetha097) [reply rocAL_pybind/rocal_pybind.cpp]: removed different functions for gpy vs the cpu case. Single function handles both

**Comment by swetha097** [comment]
> requested few changes

Resolved the comments

**Review by LakshmiKumar23 — CHANGES_REQUESTED** [review]
@swetha097 please add a test case to this in the shell script. Also please add a comment on the PR on how to test this case alone. I tried to figure it out but there are way too many parameters and I cannot pass the reader_type.

File: rocAL_pybind/amd/rocal/pipeline.py [hunk 6]
```diff
@@ -26,7 +26,6 @@
 import rocal_pybind as b
```
Comment (rrawther) [inline rocAL_pybind/amd/rocal/pipeline.py:26]: in another PR, we should rename rocal_pybind to more meaningful name like rocal_backend or something.

File: rocAL_pybind/amd/rocal/plugin/pytorch.py [hunk 7]
```diff
@@ -122,14 +122,14 @@ def __next__(self):
                     output = torch.empty(
                         dimensions, dtype=getattr(torch, torch_dtype))
                     self.labels_tensor = torch.empty(
-                        self.labels_size, dtype=getattr(torch, torch_dtype))
+                        self.labels_size, dtype=getattr(torch, "int32"))
```
Comment (rrawther) [inline rocAL_pybind/amd/rocal/plugin/pytorch.py:125]: the labels  data_type is hardcoded to be int32 here. But in get_one_hot_labels() it is copied as output_mem_type. Is there a case where they mismatch? It is generally not a good idea to hardcode the data_type

**Review by rrawther — CHANGES_REQUESTED** [review]
Added some comments

File: rocAL_pybind/amd/rocal/plugin/pytorch.py (hunk 7 above)
Comment (swetha097) [reply rocAL_pybind/amd/rocal/plugin/pytorch.py:125]: The output_tensor_lists dtype can by of any datatype specified by the user
output_mem_type can be HOST / GPU.
Internally in rocAL, the data-type of the labels is always int. So we do a copy to the int32 type and later conversions are handled in python

File: rocAL_pybind/examples/rocAL_api_tf_classification_reader.py [hunk 8]
```diff
@@ -48,7 +48,7 @@ def main():
     rocal_cpu = False if args.rocal_gpu else True
     device = "cpu" if rocal_cpu else "gpu"
     batch_size = args.batch_size
-    one_hot_labels = 0
+    one_hot_labels = 1
```
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL_pybind/examples/rocAL_api_tf_classification_reader.py]: Added as an option in parse_args

**Comment by swetha097** [comment]
> @swetha097 please add a test case to this in the shell script. Also please add a comment on the PR on how to test this case alone. I tried to figure it out but there are way too many parameters and I cannot pass the reader_type.

Added a test case in PYTHON_UNIT_TEST where we test with a reference list

**Review by LakshmiKumar23 — APPROVED** [review]
Tested on Navi21 with one hot.

**Review by rrawther — APPROVED** [review]

**Pull request merged**